[Style] Convert the 'scroll-snap-align' and 'scroll-snap-type` properties to strong style types

weinig opened the pull request:
#### 37cb1c56ae303478119a1a9afcfb44634661f9c4
<pre>
[Style] Convert the &apos;scroll-snap-align&apos; and &apos;scroll-snap-type` properties to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=299259">https://bugs.webkit.org/show_bug.cgi?id=299259</a>

Reviewed by Darin Adler.

Converts the &apos;scroll-snap-align&apos; and &apos;scroll-snap-type` properties
to use strong style types.

To simplify usage of &quot;minimally serializing / coalescing&quot; tuples, this
change adds a new modifier that any tuple can specialize to specify
if it wants the &quot;minimally serializing / coalescing&quot; behavior. This is
parallel to the existing modifier that allows specifying what kind of
separator to use for tuples and ranges. With this, all the existing
duplicate specializations for the MinimallySerializing* types can be
removed, as the main tuple specializations now handle it.

The Style::ScrollSnapAlign type added in this change makes use of this
new functionality, marking itself as &quot;minimally serializing / coalescing&quot;.

Also fixes the &apos;scroll-snap-type&apos; parser to not allow additional keywords
after an initial &quot;none&quot; keyword.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/parsing/scroll-snap-type-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/parsing/scroll-snap-type-invalid.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleScrollSnapPoints.cpp: Removed.
* Source/WebCore/rendering/style/StyleScrollSnapPoints.h: Removed.
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapAlign.cpp: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapAlign.h: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapType.cpp: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapType.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/300415@main">https://commits.webkit.org/300415@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b0aa5fd560653739c15ce54524380a476d01fbd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122481 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32874 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/129090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42907 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50782 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/129090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125433 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34220 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/109667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/129090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27824 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/72572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/103862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28035 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/131813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49422 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/37613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/131813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/49796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105888 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/131813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25752 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/25026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49280 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/55029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/48747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50429 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.cpp:
```diff
@@ -37,20 +37,24 @@ namespace CSSPropertyParserHelpers {
-    return CSSValueList::createSpaceSeparated(firstValue.releaseNonNull());
+    auto secondValue = consumeIdentRaw<CSSValueProximity, CSSValueMandatory>(range);
+    if (!secondValue || *secondValue == CSSValueProximity)
+        return CSSPrimitiveValue::create(*firstValue);
```
Maybe could write one of these? But maybe it’s less logical? But also, maybe we will be deleting this code soon?
```suggestion
    if (secondValue != CSSValueMandatory)
        return CSSPrimitiveValue::create(*firstValue);
```
```suggestion
    if (secondValue.value_or(CSSValueProximity) == CSSValueProximity)
        return CSSPrimitiveValue::create(*firstValue);
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.cpp:
```diff
@@ -37,20 +37,24 @@ namespace CSSPropertyParserHelpers {
+    if (!secondValue || *secondValue == CSSValueProximity)
+        return CSSPrimitiveValue::create(*firstValue);
+
+    return CSSValueList::createSpaceSeparated(CSSPrimitiveValue::create(*firstValue), CSSPrimitiveValue::create(*secondValue));
```
Could do this. But maybe we will be deleting this code soon enough that we shouldn’t bother?
```suggestion
   return CSSValueList::createSpaceSeparated(CSSPrimitiveValue::create(*firstValue), CSSPrimitiveValue::create(CSSValueMandatory));
```

weinig replied:
The future is so nebulous. Unless I am just about to remove it, I say fix it. So...fixed!

weinig committed:
[Style] Convert the 'scroll-snap-align' and 'scroll-snap-type` properties to strong style types
https://bugs.webkit.org/show_bug.cgi?id=299259

Reviewed by Darin Adler.

Converts the 'scroll-snap-align' and 'scroll-snap-type` properties
to use strong style types.

To simplify usage of "minimally serializing / coalescing" tuples, this
change adds a new modifier that any tuple can specialize to specify
if it wants the "minimally serializing / coalescing" behavior. This is
parallel to the existing modifier that allows specifying what kind of
separator to use for tuples and ranges. With this, all the existing
duplicate specializations for the MinimallySerializing* types can be
removed, as the main tuple specializations now handle it.

The Style::ScrollSnapAlign type added in this change makes use of this
new functionality, marking itself as "minimally serializing / coalescing".

Also fixes the 'scroll-snap-type' parser to not allow additional keywords
after an initial "none" keyword.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/parsing/scroll-snap-type-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-snap/parsing/scroll-snap-type-invalid.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleScrollSnapPoints.cpp: Removed.
* Source/WebCore/rendering/style/StyleScrollSnapPoints.h: Removed.
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapAlign.cpp: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapAlign.h: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapType.cpp: Added.
* Source/WebCore/style/values/scroll-snap/StyleScrollSnapType.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/300415@main
